src/thunderbots/software/util/parameter/generate_parameters.py

```diff
@@ -0,0 +1,291 @@
+    """Takes the information about the parameters and namespaces
+    and recursively generates the header file, returns a string to be written
+    to the file
+
+    NOTE: Both the header and the cpp have very similar structure, so they will
+    both be generated by this function
+
+    :param param_info: Contains namespace and parameter information
+        organized by filename, followed by namespaces until the parameter
+    :param cfg_name: The name of the configuration file this parameter is in
+    :param header_file_pointer: The file pointer to the dynamic_parameters.h file
+    :param cpp_file_pointer: The file pointer to the dynamic_parameters.cpp file
+    :type param_info: dict
+    :type cfg_name: str
+    :type header_file_pointer: file
+    :type cpp_file_pointer: file
+
+    :returns: The contents of the header file
+    :rtype: str
+
+    """
+    # iterate through keys
+    for key in param_info.keys():
+
+        # if type key is found, create a parameter
+        if "type" in param_info[key]:
+            parameter = param_info[key]
+
+            # write declaration to header file
+            header_file_pointer.write(
+                constants.H_PARAMETER_DECL.format(
+                    name=key,
+                    type=constants.CPP_TYPE_MAP[parameter["type"]]
+                )
+            )
+
+            # write instantiation to cpp file
+            cpp_file_pointer.write(
+                constants.CPP_PARAMETER_INSTNACE.format(
+                    name=key,
+                    path="/{}/{}".format(cfg_name, key),
+                    default=parameter["default"],
+                    type=constants.CPP_TYPE_MAP[parameter["type"]],
+                    quote="\"" if parameter["type"] in constants.QUOTE_TYPES else ""
+                )
+            )
+
+        # else setup the namespaces
+        else:
+            # start the namespace
+            header_file_pointer.write(constants.NAMESPACE_OPEN.format(name=key))
+            cpp_file_pointer.write(constants.NAMESPACE_OPEN.format(name=key))
+
+            __header_cpp_gen(
+                param_info[key], cfg_name, header_file_pointer, cpp_file_pointer)
+
+            # end the namespace
+            header_file_pointer.write(constants.NAMESPACE_CLOSE)
+            cpp_file_pointer.write(constants.NAMESPACE_CLOSE)
+
+
+def generate_server_node(param_info, output_path):
+    """This function must be called after the cfg, h and cpp files
+    have been generated. Those generated cfg files will be setup
+    as an indidual server containing the parameters to change
+
+    :param param_info: Return value of load_configuration, dict containing params
+    :param output_path: Where the cfgs should be saved to
+    :type param_info: dict
+    :type output_path: str
+
+    """
+    # reconfigure server node file
+    with open(output_path+"reconfigure_server_host.cc", 'w+') as reconfigure_server_node:
```

Comment:
What was the reasoning behind `.cc` vs `.cpp`?

Reply:
It was so that the node wasn't included with the rest of the h/cpp files. I initially had it as cpp but almost all of our stuff in the CMake lists just does /parameter/*.cpp, so it was also including the node, which had a main function it and everything was breaking.

Reply:
Hmmm tradeoff between the CMAKE executable being explicit about what files it needs, and filename ending?

I'd probably lean towards leaving the file extension as `.cpp` for consistency, and just forcing the cmake executables to be more specific about the parameter files they want.

Reply:
This problem has been since resolved after moving the node into its own folder.